File: 065/main.cpp
#include <iostream>
#include <string.h>
using namespace std;

class Solution
{
  public:
    bool isNumber(string s)
    {
        if (s.size() == 0)
            return false;
        int len = s.size();
        int i = 0;
        bool flag = false;
        while (i < len && s[i] == ' ')
            i++;
        if (i < len && (s[i] == '+' || s[i] == '-'))
            i++;
        while (i < len && s[i] >= '0' && s[i] <= '9')
        {
            flag = true;
            i++;
        }
        if (i < len && s[i] == '.')
        {
            i++;
            while (i < len && s[i] >= '0' && s[i] <= '9')
            {
                flag = true;
                i++;
            }
        }
        if (i < len && s[i] == 'e' && flag)
        {
            i++;
            flag = false;
            if (i < len && (s[i] == '+' || s[i] == '-'))
                i++;
            while (i < len && s[i] >= '0' && s[i] <= '9')
            {
                flag = true;
                i++;
            }
        }
        while (i < len && s[i] == ' ')
            i++;
        return flag && i == len;
    }
};

int main()
{
    Solution sol;
    string s = "e10";
    cout << sol.isNumber(s) << endl;
    return 0;
}

File: 065/main.js
/**
 * @param {string} s
 * @return {boolean}
 */

// 解法一
// var isNumber = function (s) {
//   const str = s.trim()
//   if (!str.length) return false
//   const num = +str
//   return !isNaN(num)
// }

// 解法二
var isNumber = function (s) {
  if (s == null || s.length === 0) return false
  const len = s.length
  let i = 0
  while (i < len && s[i] === ' ') i++
  let flag = false
  if (i < len && (s[i] === '+' || s[i] === '-')) i++
  while (i < len && (s[i] >= '0' && s[i] <= '9')) {
    flag = true
    i++
  }
  if (i < len && s[i] === '.') {
    i++
    while (i < len && (s[i] >= '0' && s[i] <= '9')) {
      flag = true
      i++
    }
  }
  if (i < len && s[i] === 'e' && flag) {
    flag = false
    i++
    if (i < len && (s[i] === '+' || s[i] === '-')) i++
    while (i < len && (s[i] >= '0' && s[i] <= '9')) {
      flag = true
      i++
    }
  }
  while (i < len && s[i] === ' ') i++
  return flag && i === len
}

console.log(isNumber('2e9 '))
console.log(isNumber('.-4'))
console.log(isNumber(' 005047e+6'))
